### ISSUE TITLE
Questions about prefer-default-export

### ISSUE THREAD

**Issue by TrevorKarjanis** [ni/javascript-styleguide#39]
We applied the ruleset to our new Angular application, and we found it a bit odd to have to add `default` to the component and service class exports which are all single class files. Does it make sense to have [prefer-default-export](https://github.com/benmosher/eslint-plugin-import/blob/v2.23.2/docs/rules/prefer-default-export.md) when we have [max-classes-per-file set to 1](https://github.com/ni/javascript-styleguide/blob/55785af7da36361729f53beef7093d09cb53add8/index.js#L87)? Is this possibly just an Angular specific experience? I'm interested in others' experiences and opinions.

**Comment by LukasKall** [comment]
I had the same experience in our TypeScript (non Angular) app. It's also a bit wired when you start writing new files. Often I started with a single export and added another export later, like a constant or an interface. To get rid of the default export, which I wanted in most cases, I had to adjust all imports as well.

**Comment by jattasNI** [comment]
We agreed in person to disable the rule. We want to also consider whether it makes sense to ban default exports entirely. @TrevorKarjanis will evaluate the resources posted in our meeting chat and make a recommendation.

**Comment by TrevorKarjanis** [comment]
TL;DR Read the proposal.

It seems the strongest argument for default exports from the [airbnb discussion](https://github.com/airbnb/javascript/issues/1365#issuecomment-293052808) is that they promote exporting one thing per file. After encountering the rule in development, however, we found ourselves asking why a default is needed when only one thing is exported? Logically, defining a default implies a non-default which is not the case for a file with a single export. Also, we didn't feel compelled to define a single export per file just because we were required to declare it the default. Despite the good intentions, they weren't effective and instead confusing. More importantly, defining one class per file is not the same as enforcing one export per file. In our case, if there is more than one export, usually types, we tend to import them as well, so the default isn't used. As a result, we saw a trend to prefer named imports naturally occur in our Angular applications.

It's also argued that renaming the export doesn't affect the imports. However, in our case renaming an export is rare, and most of the time we want the import and the export to have the same name. Requiring the `as` keyword is more explicit. An argument is made that, ["'Consistency' in a codebase doesn't mean 'you only do one kind of thing'. It means that when you do a thing, you do it in the same kind of way."](https://github.com/airbnb/javascript/issues/1365#issuecomment-351643415). While I agree with the concept, if prefer-default-export were disabled and no-default-export weren't enforced then a mixed codebase would again beg the question of the purpose of the default. Therefore, we might be compelled to disallow default exports, and we like the readability of consistent named imports likely do to the symmetrically aligned brackets. Also, it has been argued that disallowing defaults when multiple exports are unavoidable is a simpler consistency.

The creator of ESLint argues that, in his experience, default exports [are a guessing game](https://humanwhocodes.com/blog/2019/01/stop-using-default-exports-javascript-module/) and chose not to use them. This is probably improved with TypeScript and IDE tools, but similar sentiment can be found on a [RFC for GitLab](https://gitlab.com/gitlab-org/frontend/rfcs/-/issues/20) that provides a good overview.

All that considered, default exports are common among CommonJS developers. I also don't like disabling language features. However, if we're disabling prefer-default-export because we prefer named exports then enabling no-default-export makes sense given the above and our natural gravitation toward it already. Unfortunately, whether that is the case probably depends on the type of project.

For the projects I work on, I would propose the following.
**Proposal**
```javascript
/*
    The primary argument for default exports, promoting single export files, was not experienced in practice.
    Developers were not compelled to exercise this practice, and instead, found defining a default when there is only
    one export to be counterintuitive. While defining a class per file is recommended, multiple exports are not
    disallowed. It was also found that another benefit, renaming an export doesn't affect imports, is usually
    irrelevant, because the import and export names are preferred to be the same. See no-default-export for more
    information.
*/
'prefer-default-export': false,

/*
    With prefer-default-export disabled, disallow default exports and prefer named exports to enforce consistency,
    consistent names, require more explicit renaming, and reduce the effort spent refactoring. Multi-export files are
    expected, and a trend to prefer named imports was seen in practice.
*/
'no-default-export': 'error'
```

**Comment by rajsite** [comment]
More reasons to not prefer default exports: [the semantics are different from other exports](https://jakearchibald.com/2021/export-default-thing-vs-thing-as-default/). TL;DR default exports are the current value, unlike other exports which are a binding.
